libass/ass_font.c

```diff
@@ -620,3 +523,87 @@ void ass_font_clear(ASS_Font *font)
     }
     free((char *) font->desc.family.str);
 }
+
+/**
+ * \brief Convert glyph into ASS_Outline according to decoration flags
+ **/
+bool ass_get_glyph_outline(ASS_Outline *outline, int32_t *advance,
+                           FT_Face face, FT_Glyph glyph, unsigned flags)
+{
+    int32_t y_scale = face->size->metrics.y_scale;
+    int32_t adv = flags & DECO_ROTATE ? face->glyph->linearVertAdvance : glyph->advance.x;
+    adv = ((adv >> 9) + 1) >> 1;  // cannot overflow
```

Comment:
Is this an inlined version of `d16_to_d6`? I’d prefer that to be used explicitly and, if overflow is a concern, `d16_to_d6` itself to be adjusted to use this code.

Reply:
I'm not sure if `face->glyph->linearVertAdvance + 512` can overflow. `glyph->advance.x + 512` certainly cannot following mention of it in `FT_Get_Glyph()` docs. Actually, after this PR it will be the only use of `d16_to_d6()` so maybe better to remove `d16_to_d6()` then?

Reply:
My thinking is that the conversion functions make intent obvious and allow us to centralize any rounding/overflow fixes (like now). I now see that they probably apply in the pos/size/desc conversions below, too.

Reply:
Replaced with `d16_to_d6()` call.